File: src/data/ed269.ts
const PORTUGAL_INDEX='https://dnt.anac.pt/json/?';
let portugalUrl:Promise<string>|undefined;

function portugalVersionScore(href:string){
 const match=decodeURIComponent(href).match(/UASZoneVersion\D*(\d{2})(\d{2})(\d{4})(\d{2})(\d{2})(\d{2})/i);
 if(!match)return 0;
 const [,day,month,year,hour,minute,second]=match;
 return Number(`${year}${month}${day}${hour}${minute}${second}`);
}

export function latestPortugalEd269Url(){
 if(!portugalUrl)portugalUrl=fetch(PORTUGAL_INDEX).then(async response=>{
  if(!response.ok)throw new Error(`Portugal ED-269 directory failed (${response.status})`);
  const html=await response.text(),document=new DOMParser().parseFromString(html,'text/html');
  const links=[...document.querySelectorAll('a[href]')]
   .map(link=>link.getAttribute('href')??'')
   .filter(href=>/UASZoneVersion.*\.json$/i.test(decodeURIComponent(href)))
   .sort((left,right)=>portugalVersionScore(left)-portugalVersionScore(right));
  const href=links.at(-1);
  if(!href)throw new Error('Portugal ED-269 directory contained no zone file');
  return new URL(href,PORTUGAL_INDEX).href;
 }).catch(error=>{portugalUrl=undefined;throw error});
 return portugalUrl;
}

function circleGeometry(center:number[],radiusMetres:number,segments=96){
 const earthRadius=6371008.8,longitude=center[0]*Math.PI/180,latitude=center[1]*Math.PI/180;
 const angularDistance=radiusMetres/earthRadius,ring:number[][]=[];
 for(let index=0;index<segments;index++){
  const bearing=2*Math.PI*index/segments;
  const targetLatitude=Math.asin(Math.sin(latitude)*Math.cos(angularDistance)+Math.cos(latitude)*Math.sin(angularDistance)*Math.cos(bearing));
  const targetLongitude=longitude+Math.atan2(Math.sin(bearing)*Math.sin(angularDistance)*Math.cos(latitude),Math.cos(angularDistance)-Math.sin(latitude)*Math.sin(targetLatitude));
  ring.push([((targetLongitude*180/Math.PI+540)%360)-180,targetLatitude*180/Math.PI]);
 }
 ring.push(ring[0]);
 return{type:'Polygon',coordinates:[ring]};
}

export function normalizeEd269(payload:any){
 const features:any[]=[];
 for(const [zoneIndex,zone] of (payload.features??[]).entries()){
  const volumes=Array.isArray(zone.geometry)?zone.geometry:[zone.geometry];
  for(const [volumeIndex,volume] of volumes.entries()){
   const projection=volume?.horizontalProjection;
   if(!projection)continue;
   const geometry=projection.type==='Circle'
    ?circleGeometry(projection.center,Number(projection.radius))
    :projection.type==='Polygon'||projection.type==='MultiPolygon'?projection:undefined;
   if(!geometry)continue;
   const authority=(zone.zoneAuthority??[])[0]??{};
   const {geometry:_zoneGeometry,...zoneProperties}=zone;
   const {horizontalProjection:_projection,...volumeProperties}=volume;
   features.push({
    type:'Feature',
    id:`${zone.identifier??`zone-${zoneIndex}`}-${volumeIndex}`,
    properties:{...zoneProperties,...volumeProperties,sourceGeometryType:projection.type,circleCenter:projection.center,circleRadiusMetres:projection.radius,authorityName:authority.name,authorityService:authority.service,authorityEmail:authority.email,authorityPhone:authority.phone},
    geometry
   });
  }
 }
 return{type:'FeatureCollection' as const,features,title:payload.title,description:payload.description};
}

File: pipeline/adapters/ed269.py
"""Normalize EUROCAE ED-269-style UAS zone JSON to displayable GeoJSON.

The common format stores one or more vertical volumes in ``feature.geometry``.
Each volume contains a GeoJSON polygon or a metric circle under
``horizontalProjection``.  GeoJSON has no circle geometry, so circles are
approximated with geodesic polygons while the original radius and center remain
available in the feature properties.
"""
from __future__ import annotations

from copy import deepcopy
from math import asin, atan2, cos, pi, radians, sin
from typing import Any

EARTH_RADIUS_METRES = 6_371_008.8


def circle_polygon(center: list[float], radius_metres: float, segments: int = 96) -> dict:
    if len(center) < 2 or not all(isinstance(value, (int, float)) for value in center[:2]):
        raise ValueError("Circle center must contain numeric longitude and latitude")
    if not isinstance(radius_metres, (int, float)) or radius_metres <= 0:
        raise ValueError("Circle radius must be a positive number")
    if segments < 24:
        raise ValueError("Circle approximation needs at least 24 segments")

    longitude, latitude = map(radians, center[:2])
    angular_distance = float(radius_metres) / EARTH_RADIUS_METRES
    ring: list[list[float]] = []
    for index in range(segments):
        bearing = 2 * pi * index / segments
        target_latitude = asin(
            sin(latitude) * cos(angular_distance)
            + cos(latitude) * sin(angular_distance) * cos(bearing)
        )
        target_longitude = longitude + atan2(
            sin(bearing) * sin(angular_distance) * cos(latitude),
            cos(angular_distance) - sin(latitude) * sin(target_latitude),
        )
        lng = (target_longitude * 180 / pi + 540) % 360 - 180
        ring.append([round(lng, 8), round(target_latitude * 180 / pi, 8)])
    ring.append(ring[0])
    return {"type": "Polygon", "coordinates": [ring]}


def _projection(volume: dict[str, Any]) -> tuple[dict, dict]:
    projection = volume.get("horizontalProjection")
    if not isinstance(projection, dict):
        raise ValueError("ED-269 volume has no horizontalProjection")
    geometry_type = projection.get("type")
    if geometry_type in {"Polygon", "MultiPolygon"}:
        if not projection.get("coordinates"):
            raise ValueError(f"{geometry_type} projection has no coordinates")
        return deepcopy(projection), {"sourceGeometryType": geometry_type}
    if geometry_type == "Circle":
        center, radius = projection.get("center"), projection.get("radius")
        geometry = circle_polygon(center, radius)
        return geometry, {
            "sourceGeometryType": "Circle",
            "circleCenter": center,
            "circleRadiusMetres": radius,
        }
    raise ValueError(f"Unsupported ED-269 horizontal projection: {geometry_type!r}")


def normalize_ed269(payload: dict, *, attribution: str, source_url: str) -> list[dict]:
    zones = payload.get("features")
    if not isinstance(zones, list):
        raise ValueError("ED-269 response has no features array")

    features: list[dict] = []
    for zone_index, zone in enumerate(zones):
        if not isinstance(zone, dict):
            raise ValueError(f"ED-269 feature {zone_index} is not an object")
        volumes = zone.get("geometry")
        if isinstance(volumes, dict):
            volumes = [volumes]
        if not isinstance(volumes, list):
            raise ValueError(f"ED-269 feature {zone_index} has no geometry volumes")

        shared = {key: deepcopy(value) for key, value in zone.items() if key != "geometry"}
        authority = (zone.get("zoneAuthority") or [{}])[0]
        if not isinstance(authority, dict):
            authority = {}
        for volume_index, volume in enumerate(volumes):
            if not isinstance(volume, dict):
                raise ValueError(f"ED-269 feature {zone_index} volume {volume_index} is invalid")
            geometry, projection_metadata = _projection(volume)
            properties = {
                **shared,
                **{key: deepcopy(value) for key, value in volume.items() if key != "horizontalProjection"},
                **projection_metadata,
                "volumeIndex": volume_index,
                "authorityName": authority.get("name"),
                "authorityService": authority.get("service"),
                "authorityEmail": authority.get("email"),
                "authorityPhone": authority.get("phone"),
                "attribution": attribution,
                "sourceUrl": source_url,
            }
            identifier = zone.get("identifier") or f"zone-{zone_index}"
            features.append(
                {
                    "type": "Feature",
                    "id": f"{identifier}-{volume_index}",
                    "properties": properties,
                    "geometry": geometry,
                }
            )
    return features
